### PR TITLE
Keep maximum 30 candidates for scale down with drain

### PR DESCRIPTION
**Pull request by bskiba**
Issue #246

### CODE REVIEW

**Comment by bskiba** [comment]
cc: @mwielgus @MaciekPytel

File: cluster-autoscaler/core/scale_down.go [hunk 1]
```diff
@@ -55,6 +55,9 @@ const (
 	ScaleDownNodeDeleted ScaleDownResult = iota
 	// ScaleDownDisabledKey is the name of annotation marking node as not eligible for scale down.
 	ScaleDownDisabledKey = "cluster-autoscaler.kubernetes.io/scale-down-disabled"
+	// ScaleDownNonEmptyCandidatesCount is the maximum number of non empty nodes
+	// considered at once as candidates for scale down.
+	ScaleDownNonEmptyCandidatesCount = 30
```
Comment (mwielgus) [inline cluster-autoscaler/core/scale_down.go]: Make it a flag.

File: cluster-autoscaler/core/scale_down.go [hunk 2]
```diff
@@ -229,6 +251,34 @@ func (sd *ScaleDown) UpdateUnneededNodes(
 	return nil
 }
 
+// markSimulationError indicates a simulation error by clearing  relevant scale
```
Comment (mwielgus) [inline cluster-autoscaler/core/scale_down.go]: Tests please

File: cluster-autoscaler/core/scale_down.go (hunk 1 above)
Comment (MaciekPytel) [reply cluster-autoscaler/core/scale_down.go]: We already have a ton of arcane flags. Even as CA developer I generally have to look into the code to understand what they do and how to use them. And we never really test anything with their values changed.

Bottom line is I don't think anyone will ever change this and I think we should be aiming at reducing the numbers of arcane flags controlling some internal heuristics, not adding new ones.

Comment (bskiba) [reply cluster-autoscaler/core/scale_down.go]: I'm with Maciek on this one. However I name this flag and however long the description, it's going to be really hard to figure out what value to set. The only thing that kinda makes sense is what Maciek mentioned offline - add a flag to switch this off completely. This gives at least a somewhat clearer choice: better performance/responsiveness vs better scaling behavior

File: cluster-autoscaler/core/scale_down.go (hunk 2 above)
Comment (bskiba) [reply cluster-autoscaler/core/scale_down.go]: Added a test

File: cluster-autoscaler/core/scale_down.go (hunk 1 above)
Comment (mwielgus) [reply cluster-autoscaler/core/scale_down.go]: The thing is how fast can we react to any problems that are caused with the numbers picked in a wrong way. In GKE we can react much quicker if the configuration option is exposed as a flag.

Comment (bskiba) [reply cluster-autoscaler/core/scale_down.go]: Made this a flag.

File: cluster-autoscaler/core/scale_down.go [hunk 3]
```diff
@@ -41,6 +42,14 @@ import (
 	"github.com/golang/glog"
 )
 
+// ScaleDownNonEmptyCandidatesCount is the maximum number of non empty nodes
+// considered at once as candidates for scale down.
+var ScaleDownNonEmptyCandidatesCount = flag.Int("scale-down-non-empty-candidates-count", 30,
```
Comment (mwielgus) [inline cluster-autoscaler/core/scale_down.go]: We try to keep flags in main.go and have them all in AutoscalingOptions.

**Comment by mwielgus** [comment]
Check the build :).

File: cluster-autoscaler/core/scale_down.go (hunk 3 above)
Comment (bskiba) [reply cluster-autoscaler/core/scale_down.go]: Done

**Pull request merged**